python/ray/rllib/ddpg2/models.py

```diff
@@ -0,0 +1,408 @@
+
+from ray.rllib.models import ModelCatalog
+from ray.rllib.optimizers.multi_gpu_impl import TOWER_SCOPE_NAME
+
+
+def _build_s_network(registry, inputs, config):
+    frontend = ModelCatalog.get_model(registry, inputs, 1, config["model"])
+    return frontend.last_layer
+
+
+def _build_p_network(registry, inputs, dim_actions, config):
+    hiddens = config["actor_hiddens"]
+
+    action_out = inputs
+    for hidden in hiddens:
+        action_out = layers.fully_connected(
+            action_out, num_outputs=hidden, activation_fn=tf.nn.relu)
+    # The final output layer of the actor was a sigmoid layer, to bound the actions
+    # shape of action_scores is [batch_size, dim_actions]
+    action_scores = layers.fully_connected(
+        action_out, num_outputs=dim_actions, activation_fn=tf.nn.sigmoid)
+
+    return action_scores
+
+
+# As a stochastic policy for inference, but a deterministic policy for training
+# thus ignore batch_size issue when constructing a stochastic action
+def _build_action_network(
+        p_values, low_action, high_action, stochastic, eps, theta, sigma):
+    # shape is [None, dim_action]
+    deterministic_actions = (high_action-low_action) * p_values + low_action
+
+    exploration_sample = tf.get_variable(name="ornstein_uhlenbeck", dtype=tf.float32, \
+                         initializer=low_action.size*[.0], trainable=False)
+    normal_sample = tf.random_normal(shape=[low_action.size], mean=0.0, stddev=1.0)
+    exploration_value = tf.assign_add(exploration_sample, \
+                         theta * (.0-exploration_sample) + sigma * normal_sample)
+    stochastic_actions = deterministic_actions + eps * (high_action-low_action) * exploration_value
+
+    return tf.cond(
+        stochastic, lambda: stochastic_actions,
+        lambda: deterministic_actions)
+
+
+def _build_q_network(registry, state_inputs, action_inputs, config):
+    hiddens = config["critic_hiddens"]
+
+    #with tf.variable_scope("action_value"):
+    q_out = tf.concat([state_inputs, action_inputs], axis=1)
+    for hidden in hiddens:
+        q_out = layers.fully_connected(
+            q_out, num_outputs=hidden, activation_fn=tf.nn.relu)
+    q_scores = layers.fully_connected(
+        q_out, num_outputs=1, activation_fn=None)
+
+    return q_scores
+
+
+def _huber_loss(x, delta=1.0):
+    """Reference: https://en.wikipedia.org/wiki/Huber_loss"""
+    return tf.where(
+        tf.abs(x) < delta,
+        tf.square(x) * 0.5,
+        delta * (tf.abs(x) - 0.5 * delta))
+
+
+def _minimize_and_clip(optimizer, objective, var_list, clip_val=10):
+    """Minimized `objective` using `optimizer` w.r.t. variables in
+    `var_list` while ensure the norm of the gradients for each
+    variable is clipped to `clip_val`
+    """
+    gradients = optimizer.compute_gradients(objective, var_list=var_list)
+    for i, (grad, var) in enumerate(gradients):
+        if grad is not None:
+            gradients[i] = (tf.clip_by_norm(grad, clip_val), var)
```

Comment:
huber and clipping?

Reply:
As the implementation of /rllib/dqn use huber loss as well as grad clipping, I just followed it. Now, I allow users to specify whether to use huber or squared loss. Besides, the dqn hasn't expose the *delta* (i.e., the threshold of huber) in  configuration. I provide it in the configuration of DDPG now.